File: python/toolsets/event-correlator/src/src/event_correlator/engine.py
from __future__ import annotations

from datetime import datetime, timedelta
from typing import Any, Protocol

from .models import BreakEvent, PendingTrade, ReconciledEvent, TradeEvent


class CorrelationStore(Protocol):
    def ingest(
        self,
        pending: PendingTrade,
        expected_sources: list[str],
        build_audit: Any,
    ) -> tuple[bool, list[PendingTrade]]: ...

    def sweep_expired(self, now: datetime, build_audit: Any) -> dict[str, list[PendingTrade]]: ...

    def get_audit(self, trade_id: str | None, outcome: str | None, limit: int) -> list[dict[str, Any]]: ...


class Correlator:
    def __init__(
        self,
        store: CorrelationStore,
        expected_sources: list[str],
        correlation_window: timedelta,
    ) -> None:
        self._store = store
        self._expected_sources = set(expected_sources)
        self._correlation_window = correlation_window

    def ingest(self, event: TradeEvent, now: datetime) -> dict[str, Any]:
        if not event.trade_id:
            raise ValueError("trade_id is required")
        if event.source not in self._expected_sources:
            return {"status": "ignored_unknown_source", "trade_id": event.trade_id}

        pending = PendingTrade(
            trade_id=event.trade_id,
            source=event.source,
            payload=event.to_dict(),
            first_seen=now,
            deadline=now + self._correlation_window,
        )
        reconciled: ReconciledEvent | None = None

        def build_audit(events: list[PendingTrade]) -> dict[str, Any]:
            nonlocal reconciled
            reconciled = self._build_reconciled(events, now)
            return {
                "trade_id": event.trade_id,
                "outcome": "reconciled",
                "sources": reconciled.sources,
                "detail": reconciled.to_dict(),
                "occurred_at": now,
            }

        inserted, events = self._store.ingest(
            pending, sorted(self._expected_sources), build_audit
        )
        if not inserted:
            return {"status": "duplicate", "trade_id": event.trade_id}
        if not events:
            return {"status": "pending", "trade_id": event.trade_id}
        return {
            "status": "reconciled",
            "trade_id": event.trade_id,
            "reconciled": reconciled.to_dict(),
        }

    def sweep(self, now: datetime) -> list[dict[str, Any]]:
        built: dict[str, BreakEvent] = {}

        def build_audit(trade_id: str, events: list[PendingTrade]) -> dict[str, Any]:
            break_event = self._build_break(trade_id, events, now)
            built[trade_id] = break_event
            return {
                "trade_id": trade_id,
                "outcome": "break",
                "sources": break_event.received_sources,
                "detail": break_event.to_dict(),
                "occurred_at": now,
            }

        expired = self._store.sweep_expired(now, build_audit)
        return [built[trade_id].to_dict() for trade_id in sorted(expired)]

    def audit_history(
        self,
        trade_id: str | None = None,
        outcome: str | None = None,
        limit: int = 100,
    ) -> list[dict[str, Any]]:
        return self._store.get_audit(trade_id, outcome, limit)

    @staticmethod
    def _build_reconciled(events: list[PendingTrade], now: datetime) -> ReconciledEvent:
        ordered = sorted(events, key=lambda event: event.source)
        first_seen = min(event.first_seen for event in ordered)
        return ReconciledEvent(
            trade_id=ordered[0].trade_id,
            sources=[event.source for event in ordered],
            reconciled_at=now,
            match_duration_ms=int((now - first_seen).total_seconds() * 1000),
            events=[event.payload for event in ordered],
        )

    def _build_break(
        self, trade_id: str, events: list[PendingTrade], now: datetime
    ) -> BreakEvent:
        ordered = sorted(events, key=lambda event: event.source)
        received = [event.source for event in ordered]
        return BreakEvent(
            trade_id=trade_id,
            missing_sources=sorted(self._expected_sources.difference(received)),
            received_sources=received,
            detected_at=now,
            window_expiry=max(event.deadline for event in ordered),
            events=[event.payload for event in ordered],
        )

File: go/toolsets/event-correlator/src/internal/correlator/engine.go
package correlator

import (
	"context"
	"encoding/json"
	"fmt"
	"slices"
	"time"

	"github.com/solacese/event-correlator-tool/go/toolsets/event-correlator/src/internal/model"
	"github.com/solacese/event-correlator-tool/go/toolsets/event-correlator/src/internal/store"
)

// IngestOutcome describes what happened to an incoming event.
type IngestOutcome struct {
	Status     string                 `json:"status"`
	Reconciled *model.ReconciledEvent `json:"reconciled,omitempty"`
}

// Engine implements correlation rules independently of the SAM tool wrapper.
type Engine struct {
	store              store.Store
	expectedSources    map[string]struct{}
	expectedSourceList []string
	window             time.Duration
}

// NewEngine creates a correlation engine backed by the given store.
func NewEngine(s store.Store, expectedSources []string, window time.Duration) *Engine {
	expected := make(map[string]struct{}, len(expectedSources))
	for _, src := range expectedSources {
		expected[src] = struct{}{}
	}
	return &Engine{
		store:              s,
		expectedSources:    expected,
		expectedSourceList: slices.Clone(expectedSources),
		window:             window,
	}
}

// Ingest processes one trade event.
func (e *Engine) Ingest(ctx context.Context, event model.TradeEvent, now time.Time) (IngestOutcome, error) {
	if event.TradeID == "" {
		return IngestOutcome{}, fmt.Errorf("trade_id is required")
	}
	if _, known := e.expectedSources[event.Source]; !known {
		return IngestOutcome{Status: "ignored_unknown_source"}, nil
	}

	payload, err := json.Marshal(event)
	if err != nil {
		return IngestOutcome{}, fmt.Errorf("marshal event: %w", err)
	}
	pending := store.PendingTrade{
		TradeID:   event.TradeID,
		Source:    event.Source,
		Payload:   payload,
		FirstSeen: now,
		Deadline:  now.Add(e.window),
	}

	var reconciled *model.ReconciledEvent
	result, err := e.store.Ingest(ctx, pending, e.expectedSourceList, func(events []store.PendingTrade) model.AuditEntry {
		reconciled = e.buildReconciled(events, now)
		detail, _ := json.Marshal(reconciled)
		return model.AuditEntry{
			TradeID:    event.TradeID,
			Outcome:    "reconciled",
			Sources:    reconciled.Sources,
			Detail:     detail,
			OccurredAt: now,
			RecordedAt: now,
		}
	})
	if err != nil {
		return IngestOutcome{}, fmt.Errorf("ingest event: %w", err)
	}
	if !result.Inserted {
		return IngestOutcome{Status: "duplicate"}, nil
	}
	if reconciled == nil {
		return IngestOutcome{Status: "pending"}, nil
	}
	return IngestOutcome{Status: "reconciled", Reconciled: reconciled}, nil
}

// Sweep checks for trades whose correlation window has expired.
func (e *Engine) Sweep(ctx context.Context, now time.Time) ([]model.BreakEvent, error) {
	var built = make(map[string]model.BreakEvent)
	expired, err := e.store.SweepExpired(ctx, now, func(tradeID string, events []store.PendingTrade) model.AuditEntry {
		brk := e.buildBreak(tradeID, events, now)
		built[tradeID] = brk
		detail, _ := json.Marshal(brk)
		return model.AuditEntry{
			TradeID:    tradeID,
			Outcome:    "break",
			Sources:    brk.ReceivedSources,
			Detail:     detail,
			OccurredAt: now,
			RecordedAt: now,
		}
	})
	if err != nil {
		return nil, fmt.Errorf("sweep expired trades: %w", err)
	}

	breaks := make([]model.BreakEvent, 0, len(expired))
	for tradeID := range expired {
		breaks = append(breaks, built[tradeID])
	}
	slices.SortFunc(breaks, func(a, b model.BreakEvent) int { return compareStrings(a.TradeID, b.TradeID) })
	return breaks, nil
}

// AuditHistory returns persisted correlation outcomes.
func (e *Engine) AuditHistory(ctx context.Context, filter store.AuditFilter) ([]model.AuditEntry, error) {
	return e.store.GetAudit(ctx, filter)
}

func (e *Engine) buildReconciled(events []store.PendingTrade, now time.Time) *model.ReconciledEvent {
	slices.SortFunc(events, func(a, b store.PendingTrade) int { return compareStrings(a.Source, b.Source) })
	sources := make([]string, 0, len(events))
	tradeEvents := make([]model.TradeEvent, 0, len(events))
	var firstSeen time.Time

	for _, ev := range events {
		sources = append(sources, ev.Source)
		var tradeEvent model.TradeEvent
		_ = json.Unmarshal(ev.Payload, &tradeEvent)
		tradeEvents = append(tradeEvents, tradeEvent)
		if firstSeen.IsZero() || ev.FirstSeen.Before(firstSeen) {
			firstSeen = ev.FirstSeen
		}
	}

	return &model.ReconciledEvent{
		TradeID:       events[0].TradeID,
		Sources:       sources,
		ReconciledAt:  now,
		MatchDuration: now.Sub(firstSeen).Milliseconds(),
		Events:        tradeEvents,
	}
}

func (e *Engine) buildBreak(tradeID string, events []store.PendingTrade, now time.Time) model.BreakEvent {
	slices.SortFunc(events, func(a, b store.PendingTrade) int { return compareStrings(a.Source, b.Source) })
	received := make([]string, 0, len(events))
	tradeEvents := make([]model.TradeEvent, 0, len(events))
	var deadline time.Time

	for _, ev := range events {
		received = append(received, ev.Source)
		var tradeEvent model.TradeEvent
		_ = json.Unmarshal(ev.Payload, &tradeEvent)
		tradeEvents = append(tradeEvents, tradeEvent)
		if deadline.IsZero() || ev.Deadline.After(deadline) {
			deadline = ev.Deadline
		}
	}

	missing := make([]string, 0, len(e.expectedSources)-len(events))
	for src := range e.expectedSources {
		if !slices.Contains(received, src) {
			missing = append(missing, src)
		}
	}
	slices.Sort(missing)

	return model.BreakEvent{
		TradeID:         tradeID,
		MissingSources:  missing,
		ReceivedSources: received,
		DetectedAt:      now,
		WindowExpiry:    deadline,
		Events:          tradeEvents,
	}
}

func compareStrings(a, b string) int {
	if a < b {
		return -1
	}
	if a > b {
		return 1
	}
	return 0
}
